### PR TITLE
login

### PR DESCRIPTION
**Pull request by AigulTok**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -12,4 +12,6 @@ DB_URL=postgresql://dev:localpass@localhost:5400/dev
 # Log DB queries to console; Values: 1, 0
 DB_LOGS=1
 
-#REDIS_URL=redis://:localpass@localhost:6300/0
\ No newline at end of file
+#REDIS_URL=redis://:localpass@localhost:6300/0
+
+SECRET=secret
```
Comment (Winter4) [inline .env.example:17]: `SECRET` -> `JWT_SECRET`

File: source/config/default.ts [hunk 2]
```diff
@@ -6,4 +6,7 @@ export default {
   database: {
     logs: process.env.DB_LOGS,
   },
+  auth: {
+    secret: process.env.SECRET,
```
Comment (Winter4) [inline source/config/default.ts:10]: `secret` -> `jwtSecret`

File: source/api/v1/auth/redis.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { createClient } from "async-redis";
```
Comment (AigulTok) [inline source/api/v1/auth/redis.ts:1]: I have issue with redis connection. Could you check in your environment? It doesn't work with redis but work with async-redis. But still I get error in createClient: Module '"async-redis"' has no exported member 'createClient'.

Comment (Winter4) [reply source/api/v1/auth/redis.ts]: Fixed

**Review by Winter4 — COMMENTED** [review]
Well done. Check my commits and merge if it's ok

File: .env.example (hunk 1 above)
Comment (AigulTok) [reply .env.example]: 👍

File: source/config/default.ts (hunk 2 above)
Comment (AigulTok) [reply source/config/default.ts]: 👍

File: source/api/v1/login/auth.controller.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    const user = await User.getViaEmail(email);
+    if (!user) {
+      return res.status(400).json({ message: "Invalid credentials" });
+    }
+
+    const passwordMatch = await compare(password, user.password);
+    if (!passwordMatch) {
+      return res.status(400).json({ message: "Invalid credentials" });
+    }
+
```
[the remark this answers is not in the document]
Comment (AigulTok) [reply source/api/v1/login/auth.controller.ts]: 👍

File: env.d.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+declare namespace NodeJS {
```
[the remark this answers is not in the document]
Comment (AigulTok) [reply env.d.ts]: 👍

File: source/api/v1/auth/redis.ts (hunk 3 above)
Comment (AigulTok) [reply source/api/v1/auth/redis.ts]: 👍

File: .env.example [hunk 6]
```diff
@@ -12,4 +12,7 @@ DB_URL=postgresql://dev:localpass@localhost:5400/dev
 # Log DB queries to console; Values: 1, 0
 DB_LOGS=1
 
-#REDIS_URL=redis://:localpass@localhost:6300/0
\ No newline at end of file
+REDIS_URL=redis://:localpass@localhost:6300/0
+
+ACCESS_TOKEN_SECRET=access-token-secret
+REFRESH_TOKEN_SECRET=refresh-token-secret
```
[the remark this answers is not in the document]
Comment (AigulTok) [reply .env.example:18]: 👍

File: source/api/v1/auth/auth.controller.ts [hunk 7]
```diff
@@ -23,4 +31,102 @@ export default {
...
+
+  login: async function login(req: Request, res: Response) {
+    const { email, password } = req.body;
+
+    const user = await User.getViaEmail(email);
+    if (!user) {
+      return res.status(401).json({ message: "Invalid email or password" });
+    }
+
+    const passwordMatch = await compare(password, user.password);
```
[the remark this answers is not in the document]
Comment (AigulTok) [reply source/api/v1/auth/auth.controller.ts:42]: 👍

File: source/api/v1/auth/auth.controller.ts [hunk 8]
```diff
@@ -1,6 +1,14 @@
+import "dotenv/config";
```
[the remark this answers is not in the document]
Comment (AigulTok) [reply source/api/v1/auth/auth.controller.ts]: 👍

**Review by Winter4 — APPROVED** [review]

**Pull request merged**